libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+  if (collectors == nullptr || test_helpers == nullptr ||
+      info == nullptr || err == nullptr) {
+    LIBNETTEST2_EMIT_WARNING("query_bouncer: passed null pointers");
+    return false;
+  }
+  test_helpers->clear();
+  collectors->clear();
+  std::string requestbody;
+  try {
+    nlohmann::json doc;
+    doc["net-tests"] = nlohmann::json::array();
+    doc["net-tests"][0] = nlohmann::json::object();
+    doc["net-tests"][0]["input-hashes"] = nullptr;
+    doc["net-tests"][0]["name"] = nettest_name;
+    doc["net-tests"][0]["test-helpers"] = nettest_helper_names;
+    doc["net-tests"][0]["version"] = nettest_version;
+    requestbody = doc.dump();
+  } catch (const std::exception &exc) {
+    LIBNETTEST2_EMIT_WARNING("query_bouncer: cannot serialize request");
+    err->reason = 1;
+    err->library_name = "nlohmann/json";
+    err->library_version = nlohmann_json_version();
+    err->reason = exc.what();
+    return false;
+  }
+  LIBNETTEST2_EMIT_DEBUG("query_bouncer: JSON request: " << requestbody);
+  std::string responsebody;
+  // TODO(bassosimone): we should probably discuss with @hellais whether we
+  // like that currently we do not have a cloudfronted bouncer fallback. This
+  // is to be done later since I want to reach feature parity with MK legacy
+  // codebase first, and focus on perks later.
+  std::string url = without_final_slash(settings_.bouncer_base_url);
+  url += "/bouncer/net-tests";
+  LIBNETTEST2_EMIT_DEBUG("query_bouncer: URL: " << url);
+  if (!curlx_post_json(std::move(url), std::move(requestbody), curl_timeout,
+                       &responsebody, info, err)) {
+    return false;
+  }
+  LIBNETTEST2_EMIT_DEBUG("query_bouncer: JSON reply: " << responsebody);
+  try {
+    // TODO(bassosimone): make processing more flexible and robust? Here we
+    // are making strong assumptions on the returned object type. This is
+    // also something that we can defer to the future.
+    auto doc = nlohmann::json::parse(responsebody);
+    for (auto &entry : doc.at("net-tests")) {
+      {
+        EndpointInfo info;
+        info.type = endpoint_type_onion;
+        info.address = entry.at("collector");
+        collectors->push_back(std::move(info));
+      }
+      for (auto &entry : entry.at("collector-alternate")) {
+        EndpointInfo info;
+        if (entry.at("type") == "https") {
+          info.type = endpoint_type_https;
+          info.address = entry.at("address");
+        } else if (entry.at("type") == "cloudfront") {
+          info.type = endpoint_type_cloudfront;
+          info.address = entry.at("address");
+          info.front = entry.at("front");
+        } else {
+          continue;
+        }
+        collectors->push_back(std::move(info));
+      }
+#ifdef NLOHMANN_JSON_VERSION_MAJOR  // >= v3.0.0
+      for (auto &entry : entry.at("test-helpers").items()) {
+#else
+      for (auto &entry : nlohmann::json::iterator_wrapper(entry.at("test-helpers"))) {
```

Comment:
What is the reason to support two different versions . of nlohmann json?

Reply:
The fact that migrating MK to v3.0.0 is not immediately easy and integrating this library with support for v2.x allows to remove most of the places in which it's quite a headache to reliably bump the JSON library. Ideally, I'd like to drop support for v2.x as soon as it's not necessary anymore. (With the integration of this library, there is a quite large part of the MK code that can be deleted and the remainder will be easier to deal with.)

Reply:
See 3142c3c944ab9af0bb0f24aa2213b4e0e67cd401